### PR TITLE
Review on VT-ADL

### PR DESCRIPTION
**Pull request by Nicolas-Pinon**
Hi there !

A review on an anomaly detection method using transformers, CNN decoder and GMM estimation !

### CODE REVIEW

File: collections/_posts/2022-07-06-Vision_Transformer_Network_for_Image_Anomaly_Detection_and_Localization.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+#### 1.b) Encoder
+
+After the transformer, the output sequence is reduced in dimension again before being fed to the decoder. The authors used a linear projection step. They also mention a sum of the sequence, so both might be feasible and operate the same dimension reduction.
+
+$$ \mathbb{R}^{N \times D} \xrightarrow[\text{projection}]{\text{linear}} \mathbb{R}^{D}$$
+
+### 2) Decoder
+
+The decoder is a **classical CNN decoder**, consisting of 5 transposed convolutional layers, with batch-normalization and ReLu activation, except for the last layer which uses tanh activation.
```
Comment (Gasp34) [inline collections/_posts/2022-07-06-Vision_Transformer_Network_for_Image_Anomaly_Detection_and_Localization.md:58]: I think it is not classic to decode a vector (1 dim) into an image (3 dim) with convolutions, maybe there is a missing step you can precise.

File: collections/_posts/2022-07-06-Vision_Transformer_Network_for_Image_Anomaly_Detection_and_Localization.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+
+
+### 3) GMM estimation 
+
+The GMM estimation network, uses as **input** the output of the transformer (1.a), i.e. **the sequence output by the transformer**.  
+The goal is to estimate the mixture weights, means and variance of the GMM, which will be output by the network.  
+Then given the data samples and the estimated distribution, one can compute the **negative log likelihood (NLL)** of each sample.
+
+$$ \mathbb{R}^{N \times D} \xrightarrow[\text{estimation}]{\text{mixture weights}} [0, 1]^{N \times K} \times \mathbb{R_+}^{N \times K} \times \mathbb{R}^{N \times K} \xrightarrow[\text{}]{\text{NLL}} \mathbb{R}^{N}$$
```
Comment (Gasp34) [inline collections/_posts/2022-07-06-Vision_Transformer_Network_for_Image_Anomaly_Detection_and_Localization.md:70]: You should add what is K. Is it the number of gaussian in the mixture ?

File: collections/_posts/2022-07-06-Vision_Transformer_Network_for_Image_Anomaly_Detection_and_Localization.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+<sup> Note : this process is different than the "usual" process consisting for a network to output the membership of each sample, and then analytically determining the mixture weights, means and variances. </sup> 
+
+
+### Global view :
+![](/collections/images/VT-ADL/fig2.jpg)
+After inspection of the code on github we found : $$H = W = 512$$, $$P = 32$$, $$N = \frac{HW}{P^2} = 256$$, $$D = 512$$ (which is a coincidence that it is equal to $$H$$ and $$W$$).
+
+### Loss (training/testing) :
+
+The loss fonction consists of three terms to balance, two to measure the quality of the reconstruction error (the **$l_2$ norm** and the **structural similarity**) that affects the transformer, encoder and decoder, and a third term (**NLL**) that affects the transformer and the GMM estimation network.
```
Comment (Gasp34) [inline collections/_posts/2022-07-06-Vision_Transformer_Network_for_Image_Anomaly_Detection_and_Localization.md:81]:
```suggestion
The loss function consists of three terms to balance, two to measure the quality of the reconstruction error (the **$$l_2$$ norm** and the **structural similarity**) that affects the transformer, encoder and decoder, and a third term (**NLL**) that affects the transformer and the GMM estimation network.  
```

**Comment by Gasp34** [comment]
Very squared review, I made some small comments

**Review by Gasp34 — CHANGES_REQUESTED** [review]

File: collections/_posts/2022-07-06-Vision_Transformer_Network_for_Image_Anomaly_Detection_and_Localization.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+The transformer architecture is the same as in the original ViT paper. The image is decomposed in patches, embedded through a linear layer, summed with a position embedding and then passed through multiple encoding layers.
+
+$$ \mathbb{R}^{H \times W \times C} \xrightarrow[\text{division}]{\text{patch}}  \mathbb{R}^{N \times P \times P \times C} \xrightarrow[\text{layer}]{\text{linear}} \mathbb{R}^{N \times D} \xrightarrow[\text{}]{\text{transformer}} \mathbb{R}^{N \times D} $$  
+
+With $$H \times W$$ the size of the base image, $$C$$ the channel size, $$P \times P$$ the patch size and $$N$$ the number of patch.  
+
+**The reduction operation is thus done in the linear projection layer**.
+
+![](/collections/images/VT-ADL/fig2_short.jpg)
```
[the remark this answers is not in the document]
Comment (Nicolas-Pinon) [reply collections/_posts/2022-07-06-Vision_Transformer_Network_for_Image_Anomaly_Detection_and_Localization.md:45]: I think I'll leave it so everyone can remember your last review when looking at this arrow

File: collections/_posts/2022-07-06-Vision_Transformer_Network_for_Image_Anomaly_Detection_and_Localization.md (hunk 1 above)
Comment (Nicolas-Pinon) [reply collections/_posts/2022-07-06-Vision_Transformer_Network_for_Image_Anomaly_Detection_and_Localization.md:58]: added this precision

File: collections/_posts/2022-07-06-Vision_Transformer_Network_for_Image_Anomaly_Detection_and_Localization.md (hunk 2 above)
Comment (Nicolas-Pinon) [reply collections/_posts/2022-07-06-Vision_Transformer_Network_for_Image_Anomaly_Detection_and_Localization.md:70]: Precision added

**Review by Gasp34 — APPROVED** [review]

**Comment by OlivierBernard** [comment]
Excellent post !

**Pull request merged**